### DISCUSSION TITLE
Vaults: data models and UX

### DISCUSSION THREAD

**Discussion by dylanmartin** [trendscenter/coinstac-mint#14] [General]
### Vault-Hosted Dataset Integration for Studies

**Context:**  
Consortium leaders may want to include vault-hosted datasets in their study. Serving this objective effectively requires careful design of both the user interface and the data model. The following outlines a starting point for this.

---

#### UI/UX Flow:

- **Dataset First:**  
  When configuring a study that uses vault-hosted data, if a dataset is selected first, the view for computation selection is modified: Each computation is annotated to indicate whether it is compatible with the selected dataset.

- **Computation First:**  
  If a computation is selected first, each dataset in the dataset list is annotated to indicate whether it is compatible with the selected computation.

- **Vault User Integration:**  
  When a dataset is selected, the corresponding vault user is added to the consortium and configured to use the selected dataset.

---

#### Example Data Model:

```typescript
interface VaultUser {
  id: string;  
  name: string;  
  datasets: Dataset[];  
}

interface Dataset {
  id: string;  
  description: string;  
  allowedComputations: Computation[];  
}

interface Computation {
  id: string;  
  title: string;  
  imageName: string;  
  imageDownloadUrl: string;  
}
```

---

#### Implementation Considerations:

- **GraphQL Query Example:**  
  To populate the dataset list in the UI, a GraphQL query and resolver such as `getDatasets` can be implemented.

- **React State Management:**  
  The computation select and dataset select will now share some context and cause changes to one another. This needs to be handled cleanly.

---

#### Additional Concerns:

1. **Handling Different Versions of Computations:**  
   How will changes in computation versions be managed over time?

2. **Maintaining Reproducibility:**  
   How do we ensure studies remain reproducible when computation versions or datasets change?

3. **Binding Computations to Specific Images:**  
   How should computations be bound to specific container images, ensuring the correct computation version is used?

**Comment by dylanmartin** [comment]
Vaults containing public datasets are a different use case than organizations putting up private data for use in computations. There is overlap and some of the solutions may converge but lets think deeply through what we want to accomplish in each individual case before thinking about incorporating the current system implementation into the solution.

**Comment by spanta28** [comment]
Some feedback and clarity needed here:
Can both the 'Dataset first' and 'Computation first' selection be implemented if we implement the React State Management well?:
The computation select and dataset select will now share some context and cause changes to one another. This needs to be handled cleanly.

Need more discussion:
Public/Private Vault datasets:
Are private datasets only available to specific users/study? I believe this to be true, when some PI/org wants to vault their data . May need feedback from Vince. Consider : Data reconstruction 

Handling different versions of computations so that results are reproducible from Coinstac.
Can we have put some info for each computation in the meta fields like : version and brief description of the version along with its image, and may be that meta info. will be displayed in the UI and User can select that specific version for reproducing results. Always the latest version is the default one that shows up.

**Reply by dylanmartin** [reply]
To clarify the workflow described above:

The proposed workflow supports flexibility in the study configuration process, ensuring compatibility feedback is dynamically provided based on the user’s selection order. Specifically:

- **Bidirectional Workflow**:
   - If the consortium leader selects a computation first:
     - The vault selection UI should annotate the available vaults, clearly indicating which ones are compatible with the chosen computation.
   - Conversely, if a vault is selected first:
     - The computation selection UI should annotate computations to highlight compatibility with the selected vault(s).

   This approach ensures consistency and an intuitive user experience, regardless of the selection order. The shared state management between the computation and vault selections will be crucial to achieving this dynamic interaction cleanly.

**Comment by dylanmartin** [comment]
![image](https://github.com/user-attachments/assets/efc0477c-c229-4d6e-b466-c9890b306a7f)
